### PR TITLE
Raid Detection

### PR DESCRIPTION
**Pull request by saramaebee**

### CODE REVIEW

**Comment by saramaebee** [comment]
Note: while the test fails on CI, that's because it takes longer than twenty seconds. Test passes on computer, can be verified by lowering the `TIME_TILL_REMOVAL` option in config.json

File: src/config.json [hunk 1]
```diff
@@ -1,6 +1,7 @@
 {
     "COMMAND_PREFIX": "?",
-    "MEMBER_ROLE": "592088198746996768",
+	"MEMBER_ROLE": "592088198746996768",
+	"MOD_ROLE":"490594428549857281",
```
Comment (LamboCreeper) [inline src/config.json]: Formatting is not consistent here.

File: src/config.json [hunk 2]
```diff
@@ -11,12 +12,17 @@
     "PRODUCTION_ENV": "production",
     "commands_directory": "commands",
     "handlers_directory": "event/handlers",
-    "LOG_CHANNEL_ID": "405068878151024640",
+	"LOG_CHANNEL_ID": "405068878151024640",
+	"MODS_CHANNEL_ID": "495713774205141022",
```
Comment (LamboCreeper) [inline src/config.json]: Formatting is not consistent here.

File: src/event/handlers/RaidDetectionHandler.ts [hunk 3]
```diff
@@ -0,0 +1,33 @@
+import { Constants, GuildMember, TextChannel } from "discord.js";
+import EventHandler from "../../abstracts/EventHandler";
+import { MODS_CHANNEL_ID, RAID_SETTINGS, MOD_ROLE } from "../../config.json";
+
+class RaidDetectionHandler extends EventHandler {
+	constructor() {
+		super(Constants.Events.GUILD_MEMBER_ADD);
+	}
+
+	private joinQueue: GuildMember[] = [];
```
Comment (LamboCreeper) [inline src/event/handlers/RaidDetectionHandler.ts]: Can this be above the constructor please?

File: src/event/handlers/RaidDetectionHandler.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+	handle = async (member: GuildMember): Promise<void> => {
+		const timeToWait = 1000 * RAID_SETTINGS.TIME_TILL_REMOVAL;
+		const modChannel = member.guild?.channels.cache.find(channel => channel.id === MODS_CHANNEL_ID) as TextChannel;
+
+		this.joinQueue.push(member);
+
+		if (this.joinQueue.length > RAID_SETTINGS.MAX_QUEUE_SIZE) {
+			await modChannel?.send(`<@&${MOD_ROLE}>, a raid has been detected!`);
+		}
+		setTimeout(() => {
```
Comment (LamboCreeper) [inline src/event/handlers/RaidDetectionHandler.ts]: Can you add a blank line between these two, please.

File: src/event/handlers/RaidDetectionHandler.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+			const index = this.joinQueue.indexOf(member);
+
+			if (index > -1) {
```
Comment (LamboCreeper) [inline src/event/handlers/RaidDetectionHandler.ts:26]: Could you not just do this?
```suggestion
			if (this.joinQueue.includes(member)) {
```

File: test/MockDiscord.ts [hunk 6]
```diff
@@ -59,8 +59,8 @@ export default class MockDiscord {
 		return this.user;
 	}
 
-	public getGuildMember(): Discord.GuildMember {
-		return this.guildMember;
+	public getGuildMember(unique?:boolean): Discord.GuildMember {
```
Comment (LamboCreeper) [inline test/MockDiscord.ts]: Can you add a space after the `:` please.

File: test/event/handlers/RaidDetectionHandlerTest.ts [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+		it("removes member from joinQueue", done => {
+			const mockGuildMember = discordMock.getGuildMember();
+
+			handler.handle(mockGuildMember).then(() => {
+				expect(handler.joinQueue.includes(mockGuildMember)).to.be.true;
+				setTimeout(() => {
+					expect(handler.joinQueue.includes(mockGuildMember)).to.be.false;
+					done();
+				}, 1000 + RAID_SETTINGS.TIME_TILL_REMOVAL * 1000);
+			});
```
Comment (LamboCreeper) [inline test/event/handlers/RaidDetectionHandlerTest.ts]: Is there a reason you are using `.then()` over `async/await` here?

File: src/event/handlers/RaidDetectionHandler.ts [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+				this.joinQueue.splice(index, 1);
+			} else {
+				modChannel?.send("Had trouble removing user from join queue");
```
Comment (LamboCreeper) [inline src/event/handlers/RaidDetectionHandler.ts]: Might be good to specify which user?

**Review by LamboCreeper — CHANGES_REQUESTED** [review]
Looks good, just a few formatting things. We're going to need to fix the ability for this not to run in CI before we merge, though.

File: src/event/handlers/RaidDetectionHandler.ts (hunk 5 above)
Comment (saramaebee) [reply src/event/handlers/RaidDetectionHandler.ts:26]: i need the index to splice it, don't i? I suppose i could do 
```js
if (this.joinQueue.includes(member)) {
    this.joinQueue.splice(this.joinQueue.indexOf(member), 1);
}
```

File: test/event/handlers/RaidDetectionHandlerTest.ts (hunk 7 above)
Comment (saramaebee) [reply test/event/handlers/RaidDetectionHandlerTest.ts]: that was the only way to make it work. see discord bot channel

**Comment by saramaebee** [comment]
> Looks good, just a few formatting things. We're going to need to fix the ability for this not to run in CI before we merge, though.

i mean i can set the `TIME_TILL_REMOVAL` to be less than 20, youll just have to put it back up to a higher number on the prod version each time.

File: src/event/handlers/RaidDetectionHandler.ts (hunk 5 above)
Comment (LamboCreeper) [reply src/event/handlers/RaidDetectionHandler.ts:26]: Oh right, my bad - yeah, just leave it as it was then! 👍

**Comment by LamboCreeper** [comment]
> > Looks good, just a few formatting things. We're going to need to fix the ability for this not to run in CI before we merge, though.
> 
> i mean i can set the `TIME_TILL_REMOVAL` to be less than 20, youll just have to put it back up to a higher number on the prod version each time.

Hm, can we mock the value? We can't have a different config on prod to what it is in CI.

**Comment by saramaebee** [comment]
> > > Looks good, just a few formatting things. We're going to need to fix the ability for this not to run in CI before we merge, though.
> > 
> > 
> > i mean i can set the `TIME_TILL_REMOVAL` to be less than 20, youll just have to put it back up to a higher number on the prod version each time.
> 
> Hm, can we mock the value? We can't have a different config on prod to what it is in CI.

to mock the value, i could probably do some fanangling in the handle method with an optional boolean value `test`, and set it to only hold for five seconds if true?

**Comment by LamboCreeper** [comment]
> > > > Looks good, just a few formatting things. We're going to need to fix the ability for this not to run in CI before we merge, though.
> > > 
> > > 
> > > i mean i can set the `TIME_TILL_REMOVAL` to be less than 20, youll just have to put it back up to a higher number on the prod version each time.
> > 
> > 
> > Hm, can we mock the value? We can't have a different config on prod to what it is in CI.
> 
> to mock the value, i could probably do some fanangling in the handle method with an optional boolean value `test`, and set it to only hold for five seconds if true?

Oh, it's easier then that - we built a [getConfigValue()](https://github.com/codesupport/discord-bot/blob/develop/src/utils/getConfigValue.ts) method to do this - you can call that instead of `config.MY_VALUE` and easily mock it's value in a test. :)

**Comment by saramaebee** [comment]
All changes resolved

**Review by LamboCreeper — APPROVED** [review]

**Pull request merged**